src/librustc_resolve/lib.rs

```diff
@@ -3000,7 +3020,14 @@ impl<'a> Resolver<'a> {
                                             SuggestionType::NotFound => "".to_string(),
                                         }
                                     }
-                                    Field => format!("`self.{}`", path_name),
+                                    Field => {
+                                        is_field = true;
+                                        if self.is_static_method == false {
+                                            format!("`self.{}`", path_name)
+                                        } else {
+                                            "".to_string()
+                                        }
+                                    }
```

Comment:
I find this whole thing is a bit unclear. Basically we are trying to detect things based on what fields there are in the struct rather than in what kind of expression the error occurs. What you're doing, if I understand correctly, is that you emit the help message when trying to refer to a variable named `path_name` only if we are in a static method of a type that has a field of name `path_name`, wether we are in the context of a call expression or not (which is fine, the type could have a closure field... ).

But you don't catch the cases where the user would try to invoke a method from the type... Maybe this is the right thing to do since good reporting would require remembering that we are in a call-expression context, which may be a bit overkill. But I don't know... the thing is that if the type has both an field and a method named `foo`, then `foo(...)` will spawn the help message saying that you don't have access to the *fields*, but if you have only a method named `foo` and no such field, you will not have the message. All of this seems a bit inconsistent. But maybe it's not that much of a problem...